codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java

```diff
@@ -0,0 +1,265 @@
+    public boolean isActive() {
+        return !closed;
+    }
+
+    @Override
+    protected AbstractUnsafe newUnsafe() {
+        return new Unsafe();
+    }
+
+    @Override
+    protected boolean isCompatible(EventLoop loop) {
+        return true;
+    }
+
+    @Override
+    protected SocketAddress localAddress0() {
+        return parent().localAddress();
+    }
+
+    @Override
+    protected SocketAddress remoteAddress0() {
+        return parent().remoteAddress();
+    }
+
+    @Override
+    protected void doBind(SocketAddress localAddress) throws Exception { }
+
+    @Override
+    protected void doDisconnect() throws Exception {
+        closed = true;
+    }
+
+    @Override
+    protected void doClose() throws Exception {
+        closed = true;
+        if (!readInProgress) {
+            while (!inboundBuffer.isEmpty()) {
+                ReferenceCountUtil.release(inboundBuffer.poll());
+            }
+        }
+    }
+
+    @Override
+    protected void doBeginRead() throws Exception {
+        if (readInProgress) {
+            return;
+        }
+
+        if (inboundBuffer.isEmpty()) {
+            readInProgress = true;
+            return;
+        }
+
+        final InternalThreadLocalMap threadLocals = InternalThreadLocalMap.get();
+        final Integer stackDepth = threadLocals.localChannelReaderStackDepth();
+        if (stackDepth < MAX_READER_STACK_DEPTH) {
+            threadLocals.setLocalChannelReaderStackDepth(stackDepth + 1);
+            try {
+                readTask.run();
+            } finally {
+                threadLocals.setLocalChannelReaderStackDepth(stackDepth);
+            }
+        } else {
+            eventLoop().execute(readTask);
+        }
+    }
+
+    @Override
+    protected final void doWrite(ChannelOutboundBuffer in) throws Exception {
+        if (closed) {
+            throw new ClosedChannelException();
+        }
+
+        EventExecutor preferredExecutor = doWritePreferredEventExecutor();
+
+        if (preferredExecutor.inEventLoop()) {
+            for (;;) {
+                Object msg = in.current();
+                if (msg == null) {
+                    break;
+                }
+                try {
+                    doWrite(ReferenceCountUtil.retain(msg));
+                } catch (Throwable t) {
+                    // It would be nice to fail the future, but we can't do that if not on the event
+                    // loop. So we instead opt for a solution that is consistent.
+                    pipeline().fireExceptionCaught(t);
+                }
+                in.remove();
+            }
+            doWriteComplete();
+        } else {
+            // Use a copy because the original msgs will be recycled by AbstractChannel.
+            final Object[] msgsCopy = new Object[in.size()];
+            for (int i = 0; i < msgsCopy.length; i ++) {
+                msgsCopy[i] = ReferenceCountUtil.retain(in.current());
+                in.remove();
```

Comment:
@ejona86 @normanmaurer - Been a while since I dug into this. Where do we stand?  Still seems like semantics are not quite right in that writes are marked as successful (`in.remove()`) before the write event has actually occurred (when the `OneTimeTask` runs).